src/contracts/atlas/AtlasVerification.sol

```diff
@@ -270,168 +271,110 @@ contract AtlasVerification is EIP712, DAppIntegration {
-            // Current bitmap is empty, but about to be used.
-            // So increment lowest empty bitmap index to the next one
-            console.log("bitmap == 0:", bitmapIndex + 1);
-            console.log("Updating lowestEmptyBitmap from", asyncNonceBitIndex[account].LowestEmptyBitmap);
-            console.log("to", bitmapIndex + 1);
-
-            asyncNonceBitIndex[account].LowestEmptyBitmap = uint128(bitmapIndex + 1);
-        }
-
-        // TODO check if caching this shifted nonce in a var is cheaper - already done above
-        // Update the bitmap to reflect the nonce has been used
-        bitmap |= 1 << bitmapNonce;
-        nonceBitmap.bitmap = uint240(bitmap);
+        if (!async) {
+            // SEQUENTIAL NONCES
 
-        uint256 highestUsedBitmapNonce = uint256(nonceBitmap.highestUsedNonce);
-        if (bitmapNonce > highestUsedBitmapNonce) {
-            nonceBitmap.highestUsedNonce = uint8(bitmapNonce);
-        }
+            // Nonces must increase by 1 if sequential
+            if (nonce != nonceTracker.lastUsedSeqNonce + 1) return false;
 
-        // Update the nonceBitmap
-        asyncNonceBitmap[bitmapKey] = nonceBitmap;
+            // Return true here if simulation to avoid storing nonce updates
+            if (isSimulation) return true;
 
-        // Update the nonce tracker
-        return _updateNonceTracker(account, highestUsedBitmapNonce, bitmapIndex, bitmapNonce, async);
-    }
+            ++nonceTracker.lastUsedSeqNonce;
+        } else {
+            // ASYNC NONCES
 
-    function _updateNonceTracker(
-        address account,
-        uint256 highestUsedBitmapNonce,
-        uint256 bitmapIndex,
-        uint256 bitmapNonce,
-        bool async
-    )
-        internal
-        returns (bool)
-    {
-        NonceTracker memory nonceTracker = asyncNonceBitIndex[account];
+            uint256 bitmapIndex = ((nonce - 1) / 240) + 1; // +1 because highestFullBitmap initializes at 0
+            uint256 bitmapNonce = ((nonce - 1) % 240); // 1 -> 0, 240 -> 239. Needed for shifts in bitmap.
 
-        uint256 highestFullBitmap = uint256(nonceTracker.HighestFullBitmap);
-        uint256 lowestEmptyBitmap = uint256(nonceTracker.LowestEmptyBitmap);
+            bytes32 bitmapKey = keccak256(abi.encode(account, bitmapIndex));
+            NonceBitmap memory nonceBitmap = nonceBitmaps[bitmapKey];
+            uint256 bitmap = uint256(nonceBitmap.bitmap);
 
-        console.log("highestFullBitmap:", highestFullBitmap);
-        console.log("lowestEmptyBitmap:", lowestEmptyBitmap);
-
-        // Handle sequential nonce logic
-        if (!async) {
-            // TODO remove this completely if working
-            if (bitmapIndex != highestFullBitmap + 1) {
-                console.log("FALSE 1: bitmapIndex != highestFullBitmap + 1");
+            // Even if nonce rejected, increment highestFullAsyncBitmap to keep records up to date
+            if (!isSimulation && bitmap == FULL_BITMAP) {
```

Comment:
I'm concerned about the FULL_BITMAP check being the only way to increase the ``highestFullAsyncBitmap``

If a nonce is 'missed' in an otherwise full bitmap then the FULL_BITMAP check would fail.. but the ``getNextNonce`` func would still try to return a nonce equal to highest nonce + 1, which could exceed 240.